### ISSUE TITLE
Best practice for pointing at letsencrypt cert maintained by another image

### ISSUE THREAD

**Issue by andrewlow** [docker-mailserver/docker-mailserver#2321]
# Subject
I would like some feedback concerning a use case

## Description

I'm running version `10.3.0`

I use a volume map to point at the host filesystem for my let's encrypt certificate
```
 -v /path/to/config/etc/letsencrypt:/etc/letsencrypt \
```
This works well for the most part.

The let's encrypt certificate is maintained by another image - https://docs.linuxserver.io/general/swag
[as an aside, I think the [linuxserver.io ](https://fleet.linuxserver.io/) folks provide a lot of great images]

This is a nginx server with let's encrypt built in for acquiring your certificates. However, I believe this problem would happen with any other let's encrypt maintenance - possibly even if it was not in a container.

This morning I was messing around with creating a new subdomain, this would have caused the swag image to cause the let's encrypt certificate to be re-issued.  I noticed that mail was subtly busted and looked into the logs

```
Dec 11 10:26:32 mail postfix/smtpd[1766]: warning: error opening chain file: /etc/letsencrypt/live/lowtek.ca/privkey.pem: No such file or directory
```

This was easy to remedy by restarting the mail server image. However, it makes me think there is a problem with how I have this configured. 

I'm assuming that for whatever reason - the mail server was holding on to a path that had been deleted.
Even though a new path was created, the mailserver didn't re-read the path / file - and continued to reference the deleted file? Or maybe this is a docker / host file sharing problem?

Given that others will probably be also using let's encrypt for certificates I figured it was worth mentioning. Either I've done something bad in my configuration - or this is just a corner case on certificate renewal/change that might bite others.

It seems hard to believe that in the months I've been running this, I haven't had a certificate / mailserver update issue - but the certificates don't update all that often, and my server is configured to automatically reboot for security patches - so it will reboot weekly in general - an uptime of 30 days is rare. So it is possible this is a bug that will catch others with the right timing - and regular reboots/restarts avoids it.

**Comment by casperklein** [comment]
If you use `SSL_TYPE=manual`, you don't depend on any DMS logic for detecting the correct path/certificate. Could that be an option for you?

**Comment by andrewlow** [comment]
My `SSL_TYPE` setting is not manual

```
$ grep SSL_TYPE mailserver.env 
SSL_TYPE=letsencrypt
```

and based on the [doc](https://docker-mailserver.github.io/docker-mailserver/edge/config/security/ssl/#lets-encrypt-recommended) - it seems like for my setup this is the right option 

In general terms - the detection and use of my letsencrypt certificate is just fine - this is not a persistent error. The problem has only happened once, and it was when I was modifying my letsencrypt configuration in the SWAG container to add a new sub-domain (and thus forcing a new certificate to be issued).

**Comment by polarathene** [comment]
> `/etc/letsencrypt/live/lowtek.ca/privkey.pem`

There was changes to `letsencrypt` handling (and cert management in general) to address some issues and a bit of refactoring for `v10.3.0`, it's possible I might have slipped up :man_shrugging: 

https://github.com/docker-mailserver/docker-mailserver/blob/c568e09e60264ec8598457f473607d57957876e3/target/scripts/startup/setup-stack.sh#L872-L899

There's the auto-discovery logic. `check-for-changes.sh` also monitors the letsencrypt folder for several possible locations derived from `hostname`/`domainname` and ENV that could have the certs it's looking for.

Mounting a directory should be fine, as file changes should propagate, whereas mounting explicit files I think retains the old inode (_if filepath is the same but inode is different in that case, then Docker continues to reference the old replaced file still, unaware of the new version_).

If you enable `DMS_DEBUG` with reproduction test case, that should help figure out what's going on better I think, or looking at the `check-for-changes.sh` service log (_which is separate from default logs output IIRC_). Both of those had additional logging added for `v10.3.0` to improve the `letsencrypt` test cases (_which may also be helpful to reference_), I hope to find some time to add provisioning tests with a localhost ACME service which would allow for users or maintainers to better reproduce and troubleshoot these sorts of things.

---

I'm not too familiar with symlinks or how they're handled with docker bind volumes, but if your SWAG container is using those like the `certbot` container apparently does, you'll note the docs mention symlink usage requires mounting more than the cert folder alone, as it also tries to reference another archive folder that stores the actual certs.

It may be that the symlink itself never changed, but the underlying file contents did? From what I've heard that should still work fine though :man_shrugging: 

> I'm assuming that for whatever reason - the mail server was holding on to a path that had been deleted.

> Even though a new path was created, the mailserver didn't re-read the path / file - and continued to reference the deleted file?

Besides the initial setup at container startup, only `check-for-changes.sh` reacts to updating certs.. however it will fail if the filepath has changed (_eg: you switch from a specific domain to a wildcard one, that would otherwise match fine on container restart and work_).

That's always been the case, the `check-for-changes,.sh` needs to additionally run the startup script logic that also updates Postfix and Dovecot configs to the new cert paths to apply that, otherwise the services restart with the old filepath which for 99% of setups would be unchanged (_I found this issue while updating test cases for letsencrypt_).

I do intend to address that with a fix, but my backlog is rather busy atm and I lack the time to work on DMS until later this month. If that's something you need, let me know and I can advise if you want to contribute the improvement, otherwise it should be in `edge` some time in Jan.

---

Additional note, presently only `letsencrypt` supports renewals IIRC, so the other modes won't be too helpful until `letsencrypt` is deprecated in favor of a "auto-discover" default logic when no explicit certs are configured via ENV.

The intention is to unify all the modes via a refactor so they can all have feature parity and simplify the TLS logic. That'll come after additional provision tests are added.

Currently `letsencrypt` mode uses the certs via the mounted volume directory that it finds, while all other modes copy to an internal path instead of using the user path. `letsencrypt` could probably copy the chosen certs to the internal path too (or symlink?) and that would avoid the need to update the Postfix and Dovecot configs if the discovered path may have changed AFAIK.

**Comment by andrewlow** [comment]
ack - I'll try to put some time aside this week to dig into reproducing and captured logs.

**Comment by andrewlow** [comment]
Ok - I can reproduce. 

Forcing a certificate renew (by adding or removing a sub-domain) will trigger the problem when sending mail.

1. start DMS
2. Modify my swag image (nginx + letsencrypt) config to change the subdomains for my certificate
3. Restart swag
4. try to send email - boom
5. restart DMS - fixed

I looked - and the `check-for-changes.sh` script https://github.com/docker-mailserver/docker-mailserver/blob/f2ac78d1548bc310ab5ceaa46750964aa6f02e14/target/supervisor/conf.d/supervisor-app.conf#L118
Uses /var/log/supervisor/changedetector.log

My log looks like this
```
# cat changedetector.log 
[[ TASKS ]]  2021-12-14 17:06:06 Start check-for-changes script.
[[  INF  ]]  2021-12-14 17:06:06 Using postmaster address postmaster@lowtek.ca
[[  INF  ]]  2021-12-14 17:06:16 check-for-changes is ready
[[ TASKS ]]  2021-12-14 17:13:29 Start check-for-changes script.
[[  INF  ]]  2021-12-14 17:13:29 Using postmaster address postmaster@lowtek.ca
[[  INF  ]]  2021-12-14 17:13:39 check-for-changes is ready
```

The first restart with 17:06 - I was adding the debug env var `DMS_DEBUG=1` - here is confirmation it is set
```
root@mail:/var/log/supervisor# env | grep DEBUG
DMS_DEBUG=1
```

The second restart was at 17:13 - see step 5 above. I wanted to fix my mailserver that was busted.

**Comment by polarathene** [comment]
I can investigate this later today possibly. If you'd like to beat me to it, I think you could get away with using or creating your own certs locally as [documented/provided here](https://github.com/docker-mailserver/docker-mailserver/tree/master/test/test-files/ssl/example.test).

These should work fine with `SSL_TYPE=letsencrypt`, in fact the test suite for that mode does use them.. but I don't believe we actually test mail delivery, just that the SSL/TLS cert was configured/updated appropriately.

---

Just to confirm, the file path never changes when you add/remove the subdomain to the cert provisioned? It remains as `/etc/letsencrypt/live/lowtek.ca/privkey.pem` + `/etc/letsencrypt/live/lowtek.ca/fullchain.pem` before and after right?

**Comment by andrewlow** [comment]
@polarathene - I was just looking further into this. I now understand your comment about [check-for-changes.sh](https://github.com/docker-mailserver/docker-mailserver/blob/5254f7c658bfaf781d62c6ad687b455deab85c0e/target/scripts/check-for-changes.sh#L80) and Traefik. 

And while sure I can probably hack up a solution - right now it's manageable via manual intervention by just restarting containers. Ideally I'm hoping that we head off a problem for others.

> Just to confirm, the file path never changes when you add/remove the subdomain to the cert provisioned? It remains as /etc/letsencrypt/live/lowtek.ca/privkey.pem + /etc/letsencrypt/live/lowtek.ca/fullchain.pem before and after right?

Correct - the path stays the same - and a DMS container restart fixes me right up. 

I _know_ this happens when adding/removing sub-domains. It is possible that this problem will not happen on a simple certificate renewal.

For reference - this is the swag image: https://github.com/linuxserver/docker-swag
It appears to pull in certbot https://pkgs.alpinelinux.org/package/edge/community/x86_64/certbot

Hmm.. maybe we should consider more smarts around letsencrypt certificate renewals? It seems it's normal that after certbot updates things to require postfix to restart? https://community.letsencrypt.org/t/certbot-dovecot-postfix-certificate-renewal-issue/72226/4

The error feels like postfix isn't holding the key in memory, but handing on to the filehandle
```
Dec 14 17:12:29 mail postfix/smtps/smtpd[5688]: warning: error opening chain file: /etc/letsencrypt/live/lowtek.ca/privkey.pem: No such file or directory
```

**Comment by andrewlow** [comment]
Perhaps the most straight forward integration with certbot might be to copy the certs - then have a process that looks for changes. When there is a change, copy it again and restart postfix.

**Comment by polarathene** [comment]
> Perhaps the most straight forward integration with certbot might be to copy the certs - then have a process that looks for changes. When there is a change, copy it again and restart postfix.

I mentioned in my original response that `check-for-changes.sh` may only be handling Traefik `acme.json` renewals, rather than non-Traefik cert files with `SSL_TYPE=letsencrypt`, I know the other modes don't support runtime renewals currently (_this is all code that existed before I joined and I've only recently become familiar with and started refactoring with the current release_).

Postfix does restart with change detection events, and other modes do make an internal copy, however that wouldn't help currently unless updating that copy as well. I can look over `check-for-changes.sh` later today, but I suspect that if the file path remains unchanged, Postfix being restarted should switch to the correct cert without issue, only if Postfix wasn't being restarted should it be an issue.

---

**UPDATE:** While investigating this, I have found that the checksum method is broken for the cert files.. due to [this line](https://github.com/docker-mailserver/docker-mailserver/blob/5254f7c658bfaf781d62c6ad687b455deab85c0e/target/scripts/helper-functions.sh#L251):

https://github.com/docker-mailserver/docker-mailserver/blob/5254f7c658bfaf781d62c6ad687b455deab85c0e/target/scripts/helper-functions.sh#L251

Basically the quotes treat all file paths it stores as a single string which the command doesn't seem to like.

I still need to verify and reproduce on my end, but I think that's the problem here, preventing an entry in the checksum file.

**Comment by polarathene** [comment]
> `Dec 14 17:12:29 mail postfix/smtps/smtpd[5688]: warning: error opening chain file: /etc/letsencrypt/live/lowtek.ca/privkey.pem: No such file or directory`

I've not been able to reproduce this log, although I haven't used the SWAG container, but my own generated certs as per the docs I linked - This should be roughly equivalent, other than file management.

The only way I got this specific log output was by renaming the cert file `docker-mailserver` was using while running and then sending an email with TLS. So I'm going to re-emphasize that the cert management by the container you use, may (_like `certbot`_) use symlinks with file storage elsewhere, or the file paths in the provisioned folder are no longer the same.

In my tests I had no problem altering the certificate to adjust the SANs, so long as one of those SAN was the expected FQDN that `docker-mailserver` handled. In my case I used `hostname: mail.example.test` with an email account of `hello@example.test`, the MX record was to `mail.example.test`, but email was sent to `example.test` FQDN (_which for whatever reason we implicitly configure Postfix for by truncating the short hostname, the `mail` DNS label_).

If my certificate only had `mail.example.test` it would fail to establish the TLS connection (via starttls enforced on port 25). I added some other SAN with random subdomains, these did not impact the outcome, nor did the Postfix service seem to require a restart.

---

`check-for-changes.sh` won't notice these cert changes due to the bug discovered in my previous comment, so thanks for raising the issue!

I'm not sure that this would resolve the actual problem you're facing though based on my own experiences, you'll want to investigate the file path and symlink concerns on your end regarding that.

---

During my testing, I used the following, which isn't enough for you to reproduce, but should give some insights:

```bash
# Create a root CA and leaf certs, can later re-run the leaf cert command with different SAN values and run cp again to overwrite previous cert given to Postfix

mkdir tls && cd tls

step certificate create "Smallstep Root CA" ca-cert.ecdsa.pem ca-key.ecdsa.pem \
  --no-password --insecure \
  --profile root-ca \
  --not-before "2021-01-01T00:00:00+00:00" \
  --not-after "2031-01-01T00:00:00+00:00" \
  --kty EC --crv P-256

step certificate create "Smallstep Leaf" cert.ecdsa.pem key.ecdsa.pem \
  --no-password --insecure \
  --profile leaf \
  --ca ca-cert.ecdsa.pem \
  --ca-key ca-key.ecdsa.pem \
  --not-before "2021-01-01T00:00:00+00:00" \
  --not-after "2031-01-01T00:00:00+00:00" \
  --san "mail.example.test" \
  --san "random.example.test" \
  --san "example.test" \
  --kty EC --crv P-256

cd ..

# Make a copy with names/location structure like the `SSL_TYPE=letsencrypt` expects for the DMS container hostname:
cp tls/cert.ecdsa.pem letsencrypt/live/mail.example.test/fullchain.pem && cp tls/key.ecdsa.pem letsencrypt/live/mail.example.test/privkey.pem

# Using CoreDNS (built locally with some plugins) as local DNS resolver for MX and SPF record validation, allows for better reproducing a deployment, without any whitelisting to Postfix:
docker run --rm -itd --name coredns-test -v "${PWD}/coredns/Corefile:/Corefile:ro" cdns

# The mail server configured for letsencrypt mode and using a local DNS resolver instead of public
docker run --rm -it --name "dms" \
  --volume "${PWD}/config:/tmp/docker-mailserver" \
  --volume "${PWD}/letsencrypt:/etc/letsencrypt" \
  --env "SSL_TYPE=letsencrypt" \
  --env "DMS_DEBUG=1" \
  --hostname "mail.example.test" \
  --dns "172.17.0.3" \
  "mailserver/docker-mailserver:10.3.0"

# Small program I wrote and containerized to send a test email, in this case over port 25 with STARTTLS enforced,
# and the root CA cert (the OS usually provides this, including for letsencrypt CA) provided to verify the server leaf cert (equivalent to what letsencrypt provisions)
docker run --rm -t --name "smtp_client" \
  --volume "${PWD}/smtp-test/mail_template.toml:/mail_template.toml" \
  --volume "${PWD}/tls:/tls:ro" \
  --dns "172.17.0.3" \
  smtp-test

# Previous command will succeed if all goes well, otherwise fail with some error

# Renaming the cert files will cause Postfix to fail TLS negotiation, even without an internal service restart, logging the chain file error:
mv letsencrypt/live/mail.example.test/fullchain.pem letsencrypt/live/mail.example.test/fullchainx.pem && mv letsencrypt/live/mail.example.test/privkey.pem letsencrypt/live/mail.example.test/privkeyx.pem
```

**Comment by polarathene** [comment]
A potential fix was merged, it should become available in the `edge` image, otherwise `10.4.0` which should be released in roughly a week.

If that doesn't resolve the issue for you, it may be specific to your provisioning setup.
